### PR TITLE
Mm 16292 support command overrides

### PR DESCRIPTION
**Pull request by crspeller**
#### Summary
Support command overrides. Built in commands can now be overridden by plugins and custom commands.
Also fixes a bug in slash command autocomplete where duplicate checking was done incorrectly. 

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-26201

### CODE REVIEW

File: app/command.go [hunk 1]
```diff
@@ -77,32 +77,33 @@ func (a *App) CreateCommandPost(post *model.Post, teamId string, response *model
...
+			seen[cmd.Trigger] = true
+			commands = append(commands, cmd)
+		}
+	}
+
 	for _, value := range commandProviders {
 		if cmd := value.GetCommand(a, T); cmd != nil {
 			cpy := *cmd
-			if cpy.AutoComplete && !seen[cpy.Id] {
+			if cpy.AutoComplete && !seen[cpy.Trigger] {
```
Comment (marianunez) [inline app/command.go:91]: We should log here that a built-in command has been overriden.

File: app/command.go [hunk 2]
```diff
@@ -77,32 +77,33 @@ func (a *App) CreateCommandPost(post *model.Post, teamId string, response *model
...
-
 	if *a.Config().ServiceSettings.EnableCommands {
 		teamCmds, err := a.Srv().Store.Command().GetByTeam(teamId)
 		if err != nil {
 			return nil, err
 		}
 
 		for _, cmd := range teamCmds {
-			if cmd.AutoComplete && !seen[cmd.Id] {
+			if cmd.AutoComplete && !seen[cmd.Trigger] {
```
Comment (marianunez) [inline app/command.go:106]: From your PR description, I understood that you also want to allow custom commands to be able to override as well. 
Should we iterate on these before or remove these conditions then ?

File: app/plugin_commands_test.go [hunk 3]
```diff
@@ -217,4 +217,89 @@ func TestPluginCommand(t *testing.T) {
...
+			}
+
+			type MyPlugin struct {
+				plugin.MattermostPlugin
+
+				configuration configuration
+			}
+
+			func (p *MyPlugin) OnConfigurationChange() error {
+				p.API.LogError("hello")
```
Comment (marianunez) [inline app/plugin_commands_test.go:247]: Debugging log?

File: app/plugin_commands_test.go [hunk 4]
```diff
@@ -217,4 +217,89 @@ func TestPluginCommand(t *testing.T) {
...
+				if err := p.API.LoadPluginConfiguration(&p.configuration); err != nil {
+					return err
+				}
+
+				return nil
+			}
+
+			func (p *MyPlugin) OnActivate() error {
+				p.API.LogError("team", "team", p.configuration.TeamId)
```
Comment (marianunez) [inline app/plugin_commands_test.go:256]: Debugging log?

File: app/plugin_commands_test.go [hunk 5]
```diff
@@ -217,4 +217,89 @@ func TestPluginCommand(t *testing.T) {
...
+		require.Equal(t, "text", resp.Text)
+
+		err2 := th.App.DisablePlugin(pluginIds[0])
+		require.Nil(t, err2)
+
+		commands, err3 := th.App.ListAutocompleteCommands(args.TeamId, utils.T)
+		require.Nil(t, err3)
+
+		for _, commands := range commands {
+			require.NotEqual(t, "plugin", commands.Trigger)
```
Comment (marianunez) [inline app/plugin_commands_test.go:300]: I don't understand what this is asserting here given that the command registered was `/code`

File: app/command.go (hunk 1 above)
Comment (crspeller) [reply app/command.go:91]: Why? It would spam the logs for every autocomplete.

File: app/command.go (hunk 2 above)
Comment (crspeller) [reply app/command.go:106]: Yes. Fixed.

File: app/plugin_commands_test.go (hunk 3 above)
Comment (crspeller) [reply app/plugin_commands_test.go]: Ya. Just copy paste. Fixed.

File: app/plugin_commands_test.go (hunk 4 above)
Comment (crspeller) [reply app/plugin_commands_test.go]: Ya. Just copy paste. Fixed.

File: app/plugin_commands_test.go (hunk 5 above)
Comment (crspeller) [reply app/plugin_commands_test.go]: Ya. Just copy paste. Fixed.

File: app/command.go (hunk 1 above)
Comment (marianunez) [reply app/command.go]: Oh, I see. I was under the impression this was just called at the time of registering the command.

**Pull request merged**